### PR TITLE
Us004 load events faisal

### PR DESCRIPTION
**Pull request by muhamf**
Please review code for the following task for US004: Load Events

GET /collectionexercises/{collexId}/events/{eventTag} --> Retrieve event
PUT /collectionexercises/{collexId}/events/{eventTag} --> Update the event date
DELETE /collectionexercises/{collexId}/events/{eventTag} --> Delete event

I have changed in the following three classes
CollectionExerciseEndpoint.java
EventService.java
EventServiceImpl.java

### CODE REVIEW

**Comment by andrewmil** [comment]
Could we update the Api.md file to include the new endpoints? It'd be helpful for the UI stories using these

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpoint.java [hunk 1]
```diff
@@ -528,5 +530,77 @@ private CollectionExerciseDTO addCaseTypesandSurveyId(CollectionExercise collect
...
+          final @RequestBody String date)
+          throws CTPException {
+
+    Date finalDate = null;
+
+    log.info("Adding collection exercise {}, setting date to {}", id, date);
+
+      try
+        {
+          LocalDateTime parseDate = LocalDateTime.parse(date, DateTimeFormatter.ofPattern("uuuu-MM-dd'T'HH:mm:ss.SSSX", Locale.ROOT));
```
Comment (benjefferies) [inline src/main/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpoint.java]: looks like the pattern is an iso 8601 datetime. Would https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html#ISO_LOCAL_DATE_TIME work?

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpoint.java [hunk 2]
```diff
@@ -528,5 +530,77 @@ private CollectionExerciseDTO addCaseTypesandSurveyId(CollectionExercise collect
...
+  @RequestMapping(value = "/{id}/events/{tag}", method = RequestMethod.PUT, consumes = "text/plain")
+  public ResponseEntity<?> updateEventDate(
+          @PathVariable("id") final UUID id,
+          @PathVariable("tag") final String tag,
+          final @RequestBody String date)
+          throws CTPException {
+
+    Date finalDate = null;
+
+    log.info("Adding collection exercise {}, setting date to {}", id, date);
```
Comment (benjefferies) [inline src/main/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpoint.java]: Should the log be Adding collection exercise event?

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventServiceImpl.java [hunk 3]
```diff
@@ -61,4 +61,87 @@ public Event createEvent(EventDTO eventDto) throws CTPException {
...
+        {
+            Event event = this.eventRepository.findOneByCollectionExerciseAndTag(collex, tag);
+            if (event != null)
+            {
+                event.setTimestamp(new Timestamp(date.getTime()));
+
+                this.eventRepository.save(event);
+            }
+            else
+                {
```
Comment (benjefferies) [inline src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventServiceImpl.java]: formatting looks out

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventServiceImpl.java [hunk 4]
```diff
@@ -61,4 +61,87 @@ public Event createEvent(EventDTO eventDto) throws CTPException {
...
+                   throw new CTPException(CTPException.Fault.RESOURCE_NOT_FOUND, "message");
```
Comment (benjefferies) [inline src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventServiceImpl.java]: should probably add a sensible message

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventServiceImpl.java [hunk 5]
```diff
@@ -61,4 +61,87 @@ public Event createEvent(EventDTO eventDto) throws CTPException {
...
+                }
+
+                return event;
+        }
+        else
+            {
+                  throw new CTPException(CTPException.Fault.BAD_REQUEST, "message");
```
Comment (benjefferies) [inline src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventServiceImpl.java]: should probably add a sensible message

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventServiceImpl.java [hunk 6]
```diff
@@ -61,4 +61,87 @@ public Event createEvent(EventDTO eventDto) throws CTPException {
...
+        if (collex != null)
+        {
+            Event event = this.eventRepository.findOneByCollectionExerciseAndTag(collex, tag);
+            if (event != null)
+            {
+                return event;
+            }
+            else
+            {
+                throw new CTPException(CTPException.Fault.RESOURCE_NOT_FOUND, "message");
```
Comment (benjefferies) [inline src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventServiceImpl.java]: should probably add a sensible message

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventServiceImpl.java [hunk 7]
```diff
@@ -61,4 +61,87 @@ public Event createEvent(EventDTO eventDto) throws CTPException {
...
+            }
+
+        }
+        else
+        {
+            throw new CTPException(CTPException.Fault.BAD_REQUEST, "message");
```
Comment (benjefferies) [inline src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventServiceImpl.java]: should probably add a sensible message

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventServiceImpl.java [hunk 8]
```diff
@@ -61,4 +61,87 @@ public Event createEvent(EventDTO eventDto) throws CTPException {
...
+            if (event != null)
+            {
+                event.setDeleted(true);
+                this.eventRepository.delete(event);
+
+                return event;
+            }
+            else
+            {
+                throw new CTPException(CTPException.Fault.RESOURCE_NOT_FOUND, "message");
```
Comment (benjefferies) [inline src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventServiceImpl.java]: should probably add a sensible message

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventServiceImpl.java [hunk 9]
```diff
@@ -61,4 +61,87 @@ public Event createEvent(EventDTO eventDto) throws CTPException {
...
+            }
+
+        }
+        else
+        {
+            throw new CTPException(CTPException.Fault.BAD_REQUEST, "message");
```
Comment (benjefferies) [inline src/main/java/uk/gov/ons/ctp/response/collection/exercise/service/impl/EventServiceImpl.java]: should probably add a sensible message

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpoint.java [hunk 10]
```diff
@@ -528,5 +530,77 @@ private CollectionExerciseDTO addCaseTypesandSurveyId(CollectionExercise collect
...
+
+
+  /**
+   * DELETE request to delete a collection exercise event
+   * @param id Collection exercise Id
+   * @param  tag collection exercise event tag
+   * @throws CTPException on resource not found
+   * @return the collection exercise event that was to be deleted
+   */
+  @RequestMapping(value = "/{id}/{tag}", method = RequestMethod.DELETE)
```
Comment (madeye-matt) [inline src/main/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpoint.java]: This should be "/{id}/events/{tag}" to match with the other operations

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpoint.java (hunk 1 above)
Comment (madeye-matt) [reply src/main/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpoint.java]: Didn't work for me - only worked if there was a ':' in the middle of the timezone

**Pull request merged**